posthog/hogql_queries/experiments/experiment_trends_query_runner.py

```diff
@@ -226,7 +255,63 @@ def calculate(self) -> ExperimentTrendsQueryResponse:
 
         def run(query_runner: TrendsQueryRunner, result_key: str, is_parallel: bool):
             try:
-                result = query_runner.calculate()
+                database = create_hogql_database(team_id=self.team.pk)
+                if self._is_data_warehouse_query(query_runner.query):
+                    series_node = cast(DataWarehouseNode, query_runner.query.series[0])
+                    table = database.get_table(series_node.table_name)
+                    table.fields["events"] = LazyJoin(
+                        from_field=[series_node.distinct_id_field],
+                        join_table=database.get_table("events"),
+                        join_function=lambda join_to_add, context, node: (
+                            ast.JoinExpr(
+                                table=ast.SelectQuery(
+                                    select=[
+                                        ast.Alias(alias=name, expr=ast.Field(chain=["events", *chain]))
+                                        for name, chain in {
+                                            **join_to_add.fields_accessed,
+                                            "timestamp": ["timestamp"],
+                                            "distinct_id": ["distinct_id"],
+                                            "properties": ["properties"],
+                                        }.items()
+                                    ],
+                                    select_from=ast.JoinExpr(table=ast.Field(chain=["events"])),
+                                ),
+                                join_type="ASOF LEFT JOIN",
```

Comment:
I'd love to have a comment here explaining what we're doing, especially because some of how the ASOF JOIN works is only explicit in the Clickhouse docs:

```
        # ASOF JOIN finds the most recent matching event that occurred at or before each data warehouse timestamp.
        # 
        # Why this matters:
        # When a user performs an action (recorded in data warehouse), we want to know which
        # experiment variant they were assigned at that moment. The most recent $feature_flag_called
        # event before their action represents their active variant assignment.
        #
        # Example:
        #   Data Warehouse: timestamp=2024-01-03 12:00, distinct_id=user1
        #   Events: 
        #     2024-01-02: (user1, variant='control')   <- This event will be joined
        #     2024-01-03: (user1, variant='test')     <- Ignored
        #
        # This ensures we capture the correct causal relationship: which experiment variant
        # was the user assigned to when they performed the action?
```

(feel free to adjust)